### DISCUSSION TITLE
How to handle file upload on client using react hook form and next.js 14 server actions

### DISCUSSION THREAD

**Discussion by amrajat** [vercel/next.js#60575] [Help] [answered]
### Summary

I have a file input registered with react Hook form with next.js 14

```
"use client"
..........more code  
  <form action={handleSubmit(updateUserProfile)}>
        <input
                    {...register("avatar_url")}
                    type="file"
                    name="avatar_url"
                    accept="image/*"                
                  />
        <button type="submit">upload</button>
    </from>
......rest of the code goes here
```

and below is my async function

```
 async function updateUserProfile(formData) {
    // updateCurrentUserProfile is server action with supabase storage api, w/ server client
    const response = await updateCurrentUserProfile(formData);
  }
```

If I don't select any file this works smoothly without any error. but if I select a file to upload. i get error

```
Application error: a client-side exception has occurred (see the browser console for more information).
```

```
Uncaught Error: Only plain objects, and a few built-ins, can be passed to Server Actions. Classes or null prototypes are not supported.
```

I don't want to create any route handler. I want to upload the images using server actions only. I tried to construct a new FormData but that was not passing the file to the server actions instead it's just sending an empty array of FileList objects.

**Comment by icyJoseph** [comment, marked as answer]
You are mixing two things here.

Server actions are passed to the `action` prop of a form, to support progressive enhancement, and what not.

However, in your case what you want is to pass `handleSubmit(updateUserProfile)` to the `onSubmit` prop. These are very different props.

Something like:

```jsx
const onSubmit = handleSubmit(data => {
    (() => {
      action(data); // action would be your server action
    });
  });

  return (
    <form onSubmit={onSubmit}>
```

Make sure to read upon this Github issue thread: https://github.com/react-hook-form/react-hook-form/issues/10391

**Reply by amrajat** [reply]
Thank you for your answer.
